File: app/src/main/java/com/ej/rovadiahyosefcalendar/classes/SiddurKahalStringBuilder.java
package com.ej.rovadiahyosefcalendar.classes;

import android.graphics.Typeface;
import android.text.Spannable;
import android.text.SpannableStringBuilder;
import android.text.style.RelativeSizeSpan;
import android.text.style.StyleSpan;

import androidx.annotation.NonNull;

import java.util.Objects;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class SiddurKahalStringBuilder extends SpannableStringBuilder {
	public SiddurKahalStringBuilder() {
		super();
	}

	public SiddurKahalStringBuilder(Spannable kaddishText) {
		super(kaddishText);
	}

	public SiddurKahalStringBuilder(String s) {
		super(s);
	}

	public SiddurKahalStringBuilder addRepeatedKahalChorusSmallText(String inBetween, String[] stanzas, String chorus, String stanzaPrefix) {
		for (String stanza: stanzas) {
			super.append(stanzaPrefix).append(stanza);
			this.appendSmallText(chorus);

			if (!Objects.equals(inBetween, ""))
				super.append(inBetween);
		}

		if (!Objects.equals(inBetween, ""))
			super.delete(super.length() - 1, super.length());
		return this;
	}

	public SiddurKahalStringBuilder addRepeatedKahalChorusSmallText(boolean inLine, String[] stanzas, String chorus, String stanzaPrefix) {
		return this.addRepeatedKahalChorusSmallText(inLine ? "" : "\n", stanzas, chorus, stanzaPrefix);
	}

	public SiddurKahalStringBuilder addRepeatedKahalChorusSmallText(String[] stanzas, String chorus, String stanzaPrefix) {
		return this.addRepeatedKahalChorusSmallText("\n", stanzas, chorus, stanzaPrefix);
	}

	public SiddurKahalStringBuilder addRepeatedKahalChorusSmallText(String[] stanzas, String chorus) {
		return this.addRepeatedKahalChorusSmallText("\n", stanzas, chorus, "");
	}

	public SiddurKahalStringBuilder addBoldFirstLetterOfStanza(String inBetween, String[] stanzas, String stanzaPrefix) {
		for (String stanza : stanzas) {
			int stanzaStart = this.length(); // capture position before appending
			super.append(stanzaPrefix).append(stanza);

			Pattern pattern = Pattern.compile("[\u05D0-\u05EA][\u0591-\u05C7]*");
			Matcher matcher = pattern.matcher(stanza);

			if (matcher.find()) {
				int start = stanzaStart + stanzaPrefix.length() + matcher.start();
				int end = stanzaStart + stanzaPrefix.length() + matcher.end();

				this.setSpan(new StyleSpan(Typeface.BOLD), start, end, SpannableStringBuilder.SPAN_EXCLUSIVE_EXCLUSIVE);
			}

			if (!Objects.equals(inBetween, ""))
				super.append(inBetween);
		}

		if (!Objects.equals(inBetween, ""))
			super.delete(super.length() - 1, super.length());

		return this;
	}

	public SiddurKahalStringBuilder addBoldFirstLetterOfStanzaForKahal(String inBetween, String[] stanzas, String stanzaPrefix) {
		for (String stanza : stanzas) {
			int stanzaStart = this.length(); // capture position before appending
			super.append(stanzaPrefix).append(stanza);

			Pattern pattern = Pattern.compile("[\u05D0-\u05EA][\u0591-\u05C7]*");
			Matcher matcher = pattern.matcher(stanza);

			if (matcher.find()) {
				int start = stanzaStart + stanzaPrefix.length() + matcher.start();
				int end = stanzaStart + stanzaPrefix.length() + matcher.end();

				this.setSpan(new StyleSpan(Typeface.BOLD), start, end, SpannableStringBuilder.SPAN_EXCLUSIVE_EXCLUSIVE);
			}

			if (!Objects.equals(inBetween, ""))
				this.appendSmallText(inBetween);
		}

		if (!Objects.equals(inBetween, ""))
			super.delete(super.length() - 1, super.length());

		return this;
	}

	public SiddurKahalStringBuilder addBoldFirstLetterOfStanza(boolean inLine, String[] stanzas, String stanzaPrefix) {
		return this.addBoldFirstLetterOfStanza(inLine ? "" : "\n", stanzas, stanzaPrefix);
	}

	public SiddurKahalStringBuilder appendSmallText(String smallText) {
		int preSmall = super.length();
		super.append(smallText);
		super.setSpan(new RelativeSizeSpan(0.7f), preSmall, super.length(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);

		return this;
	}

	public SiddurKahalStringBuilder appendAmen() {
		return appendSmallText(" [אמן] ");
	}

	@NonNull
	@Override
	public SiddurKahalStringBuilder append(char text) {
		super.append(text);
		return this;
	}

	@NonNull
	public SiddurKahalStringBuilder append(String text) {
		super.append(text);
		return this;
	}
}

File: app/src/main/java/com/ej/rovadiahyosefcalendar/classes/SiddurKahalStringBuilder.kt
package com.ej.rovadiahyosefcalendar.classes

import android.graphics.Typeface
import android.text.SpannableStringBuilder
import android.text.style.ForegroundColorSpan
import android.text.style.RelativeSizeSpan
import android.text.style.StyleSpan
import androidx.core.text.inSpans

// This is the core builder function. It's much cleaner than a class.
fun siddurKahalBuilder(init: SiddurKahalStringBuilder.() -> Unit): CharSequence {
    return SiddurKahalStringBuilder().apply(init).build()
}

// The class is now a "context" for our DSL, containing the helper methods.
class SiddurKahalStringBuilder(
    val builder: SpannableStringBuilder = SpannableStringBuilder()
) {

    fun append(text: CharSequence): SiddurKahalStringBuilder {
        builder.append(text)
        return this
    }

    fun appendSmallText(smallText: String): SiddurKahalStringBuilder {
        builder.inSpans(RelativeSizeSpan(0.7f)) {
            append(smallText)
        }
        return this
    }

    fun appendColorText(colorText: String, color: Int): SiddurKahalStringBuilder {
        builder.inSpans(ForegroundColorSpan(color)) {
            append(colorText)
        }
        return this
    }

    fun appendAmen(): SiddurKahalStringBuilder {
        return appendSmallText(" [אמן] ")
    }

    fun addRepeatedKahalChorusSmallText(
        inBetween: String = "\n",
        stanzas: Array<String>,
        chorus: String,
        stanzaPrefix: String = ""
    ): SiddurKahalStringBuilder {
        stanzas.forEachIndexed { index, stanza ->
            append(stanzaPrefix)
            append(stanza)
            appendSmallText(chorus)

            if (inBetween.isNotEmpty() && index < stanzas.size - 1) {
                append(inBetween)
            }
        }
        return this
    }

    fun addRepeatedKahalChorusSmallText(
        inline: Boolean,
        stanzas: Array<String>,
        chorus: String,
        stanzaPrefix: String = ""
    ): SiddurKahalStringBuilder {
        return addRepeatedKahalChorusSmallText(if (inline) "" else "\n", stanzas, chorus, stanzaPrefix)
    }

    fun addRepeatedKahalChorusSmallText(
        stanzas: Array<String>,
        chorus: String,
        stanzaPrefix: String = ""
    ): SiddurKahalStringBuilder {
        return addRepeatedKahalChorusSmallText("\n", stanzas, chorus, stanzaPrefix)
    }

    fun addRepeatedKahalChorusSmallText(
        stanzas: Array<String>,
        chorus: String
    ): SiddurKahalStringBuilder {
        return addRepeatedKahalChorusSmallText("\n", stanzas, chorus, "")
    }

    fun addBoldFirstLetterOfStanza(
        inBetween: String = "\n",
        stanzas: Array<String>,
        stanzaPrefix: String = ""
    ): SiddurKahalStringBuilder {
        val hebrewLetterPattern = "[\u05D0-\u05EA][\u0591-\u05C7]*".toRegex()
        stanzas.forEachIndexed { index, stanza ->
            val fullStanza = stanzaPrefix + stanza
            val match = hebrewLetterPattern.find(fullStanza)

            if (match != null) {
                // Append the part before the bold letter
                append(fullStanza.substring(0, match.range.first))
                // Append the bold letter itself
                builder.inSpans(StyleSpan(Typeface.BOLD)) {
                    append(match.value)
                }
                // Append the rest of the stanza
                append(fullStanza.substring(match.range.last + 1))
            } else {
                append(fullStanza) // Append normally if no letter found
            }

            if (inBetween.isNotEmpty() && index < stanzas.size - 1) {
                append(inBetween)
            }
        }
        return this;
    }

    fun addBoldFirstLetterOfStanza(
        inline: Boolean,
        stanzas: Array<String>,
        stanzaPrefix: String = ""
    ): SiddurKahalStringBuilder {
        return addBoldFirstLetterOfStanza(if (inline) "" else "\n", stanzas, stanzaPrefix)
    }

    fun addBoldFirstLetterOfStanzaForKahal(
        inBetween: String = "\n",
        stanzas: Array<String>,
        stanzaPrefix: String = ""
    ): SiddurKahalStringBuilder {
        val hebrewLetterPattern = "[\u05D0-\u05EA][\u0591-\u05C7]*".toRegex()
        stanzas.forEachIndexed { index, stanza ->
            val fullStanza = stanzaPrefix + stanza
            val match = hebrewLetterPattern.find(fullStanza)

            if (match != null) {
                // Append the part before the bold letter
                append(fullStanza.substring(0, match.range.first))
                // Append the bold letter itself
                builder.inSpans(StyleSpan(Typeface.BOLD)) {
                    append(match.value)
                }
                // Append the rest of the stanza
                append(fullStanza.substring(match.range.last + 1))
            } else {
                append(fullStanza) // Append normally if no letter found
            }

            if (inBetween.isNotEmpty() && index < stanzas.size - 1) {
                appendSmallText(inBetween)
            }
        }
        return this;
    }

    // You can add your other helper methods here following the same pattern
    fun length(): Int {
        return builder.length
    }

    fun delete(start: Int, end: Int): SiddurKahalStringBuilder? {
        builder.delete(start, end)
        return this
    }

    // This method returns the final result.
    fun build(): CharSequence {
        return builder
    }

    fun clone(): SiddurKahalStringBuilder {
        // 1. Create a new SpannableStringBuilder using the constructor that
        //    is documented to copy both TEXT and SPANS from the source.
        val clonedSpannable = SpannableStringBuilder(this.builder)

        // 2. Create a new instance of our wrapper class with this perfect clone.
        return SiddurKahalStringBuilder(clonedSpannable)
    }

    fun clear() {
        builder.clear()
        builder.clearSpans()
    }
}
